tracing/tracing.go

```diff
@@ -25,10 +26,56 @@ func InstrumentHTTP(h http.Handler) http.Handler {
 		}
 
 		ctx := req.Context()
+		ctx = CtxWithTraceID(ctx, traceid)
+
 		log := slog.FromCtx(ctx)
 		log = log.With("trace_id", traceid)
 		ctx = slog.NewContext(ctx, log)
 
 		h.ServeHTTP(res, req.WithContext(ctx))
 	})
 }
+
+// CtxWithTraceID creates a new [context.Context] with the given trace ID associated with it.
+// Call [CtxGetTraceID] to retrieve the trace ID.
+func CtxWithTraceID(ctx context.Context, traceID string) context.Context {
+	return context.WithValue(ctx, traceIDKey, traceID)
+}
+
+// CtxGetTraceID gets the trace ID associated with this context.
+// Return the trace ID and true if there is a trace ID, empty and false otherwise.
+func CtxGetTraceID(ctx context.Context) (string, bool) {
```

Comment:
Instead of N functions an alternative design would be to make the keys public and the client just pass the key they want, was on the fence on this. This seemed like a more encapsulated solution (at the cost of having N functions, could get messy for a big N, but we don't plan on having a big N).

Reply:
Yep, I agree. I don't expect to have a lot of keys in our context (at least in the library). Something that I'd expect though, it's to allow users to inject their own keys, but still not sure if we have something to add here or not. 

Maybe explaining a bit more, as trace_id and org_id, we may want to allow users to inject some stuff on the logs and the context. But I'd wait a bit to come up with a solution as still not clear to me what would be the use-cases. And so far seems the log would be more important, but we already have ways to add fields there, so I think we're fine.

Reply:
On "Something that I'd expect though, it's to allow users to inject their own keys, but still not sure if we have something to add here or not." the context API itself already allows it...so not sure what do you mean :thinking: (we define a private type exactly to avoid clashing, so users could even store the trace/org ID on the context with some other key if they want, for some reason/integration/etc).

On adding things to logs, I tried going through the path of automatically extracting things from the context, I don't think that scaled well and requires custom handlers on the logger. Seems easier to just store/retrive logger instances and manage keys on the logger.

For other purposes, then we have these two functions or the user can just add any key/value they want to contexts, this functions are just helpers that will be used together with the middleware, but don't limit what people can do AFAIK.

Reply:
I meant that we have a pattern here where we want to extract an information from the header, inject it in the context, add it to the logs, and perhaps forward it to an outgoing request.

I don't think we will need much more than traceID and orgID for the moment, but I feel that application would need to have such requirements as well. UserID could be a candidate.

So, that's what I meant by expectation, I don't have enough information but perhaps we'll need to reimplement all of this for each use-case inside of different services. But again, just thinking out loud, let's discuss when we need it and IF we need it.